src/Shared/NativeMethodsShared.cs

```diff
@@ -470,6 +501,61 @@ public SystemInformationData()
             }
         }
 
+        /// <summary>
+        /// Get the exact physical core count on Windows
+        /// Useful for getting the exact core count in 32 bits processes,
+        /// as Environment.ProcessorCount has a 32-core limit in that case. 
+        /// https://github.com/dotnet/runtime/blob/221ad5b728f93489655df290c1ea52956ad8f51c/src/libraries/System.Runtime.Extensions/src/System/Environment.Windows.cs#L171-L210
+        /// </summary>
+        public unsafe static int GetPhysicalCoreCount()
+        {
+            uint len = 0;
+            const int ERROR_INSUFFICIENT_BUFFER = 122;
+
+            if (!GetLogicalProcessorInformationEx(LOGICAL_PROCESSOR_RELATIONSHIP.RelationGroup, IntPtr.Zero, ref len) &&
+                Marshal.GetLastWin32Error() == ERROR_INSUFFICIENT_BUFFER)
+            {
+                // Allocate that much space
+                var buffer = new byte[len];
+                fixed (byte* bufferPtr = buffer)
+                {
+                    // Call GetLogicalProcessorInformationEx with the allocated buffer
+                    if (GetLogicalProcessorInformationEx(LOGICAL_PROCESSOR_RELATIONSHIP.RelationGroup, (IntPtr)bufferPtr, ref len))
+                    {
+                        // Walk each SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX in the buffer, where the Size of each dictates how
+                        // much space it's consuming.  For each group relation, count the number of active processors in each of its group infos.
+                        int processorCount = 0;
+                        byte* ptr = bufferPtr;
+                        byte* endPtr = bufferPtr + len;
+                        while (ptr < endPtr)
+                        {
+                            var current = (SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX*)ptr;
+                            if (current != null && current->Relationship == LOGICAL_PROCESSOR_RELATIONSHIP.RelationGroup)
+                            {
+                                PROCESSOR_GROUP_INFO* groupInfo = &current->Group.GroupInfo;
+                                if (groupInfo != null)
+                                {
+                                    int groupCount = current->Group.ActiveGroupCount;
+                                    for (int i = 0; i < groupCount; i++)
```

Comment:
I may have misunderstood, but I thought @ladipro was thinking something more like this?

```suggestion
                                    for (int i = 0; i < groupCount && groupInfo + i < endPtr; i++)
```

Reply:
Thanks, yes, I was thinking something like this, not the `null` checks added in 4223cc2e0c875c527afaf5f67a255c2995c4ebb1. It would have to be:

```C#
for (int i = 0; i < groupCount && groupInfo + i + 1 <= endPtr; i++)
```

i.e. the entire structure pointed to by `(groupInfo + i)` fits in the buffer.

I mentioned this to @danmosemsft offline and he doesn't consider this kind of hardening necessary when calling a trusted API. @mfkl, feel free to revert the 4223cc2e0c875c527afaf5f67a255c2995c4ebb1. Apologies for randomizing you.

Reply:
Yeah, clearly we don't do it in the code this came from in dotnet/runtime and I think it would be simply "testing the OS has no bugs". But if there's a security context, don't take my word for it - we should ask others.

Reply:
All good, reverted.